src/zasm.zig

```diff
@@ -0,0 +1,716 @@
+    pub fn initUnchecked(addr: u64) PhysAddr {
+        return .{ .value = addr };
+    }
+
+    /// Creates a physical address that points to `0`.
+    pub fn zero() PhysAddr {
+        return .{ .value = 0 };
+    }
+
+    /// Convenience method for checking if a physical address is null.
+    pub fn isNull(self: PhysAddr) bool {
+        return self.value == 0;
+    }
+
+    /// Aligns the physical address upwards to the given alignment.
+    /// The alignment must be a power of 2 and greater than 0.
+    pub fn alignUp(self: PhysAddr, alignment: usize) PhysAddr {
+        return .{ .value = std.mem.alignForward(self.value, alignment) };
+    }
+
+    /// Aligns the physical address downwards to the given alignment.
+    /// The alignment must be a power of 2 and greater than 0.
+    pub fn alignDown(self: PhysAddr, alignment: usize) PhysAddr {
+        return .{ .value = std.mem.alignBackward(@intCast(usize, self.value), alignment) };
+    }
+
+    /// Checks whether the physical address has the given alignment.
+    /// The alignment must be a power of 2 and greater than 0.
+    pub fn isAligned(self: PhysAddr, alignment: usize) bool {
+        return std.mem.isAligned(self.value, alignment);
+    }
+
+    test {
+        std.testing.refAllDecls(@This());
+        try std.testing.expectEqual(@bitSizeOf(u64), @bitSizeOf(PhysAddr));
+        try std.testing.expectEqual(@sizeOf(u64), @sizeOf(PhysAddr));
+    }
+};
+
+// ~~~ [ zig-x86_64/src/stuctures/paging/page.zig ] ~~~
+
+const SIZE_4KiB_STR: []const u8 = "4KiB";
+const SIZE_2MiB_STR: []const u8 = "2MiB";
+const SIZE_1GiB_STR: []const u8 = "1GiB";
+
+pub const PageSize = enum {
+    Size4KiB,
+    Size2MiB,
+    Size1GiB,
+
+    pub fn bytes(self: PageSize) u64 {
+        return switch (self) {
+            .Size4KiB => 4096,
+            .Size2MiB => 4096 * 512,
+            .Size1GiB => 4096 * 512 * 512,
+        };
+    }
+
+    pub fn sizeString(self: PageSize) []const u8 {
+        return switch (self) {
+            .Size4KiB => SIZE_4KiB_STR,
+            .Size2MiB => SIZE_2MiB_STR,
+            .Size1GiB => SIZE_1GiB_STR,
+        };
+    }
+
+    pub fn isGiantPage(self: PageSize) bool {
+        return self == .Size1GiB;
+    }
+};
+
+// ~~~ [ zig-x86_64/src/instructions/tlb.zig ] ~~~
+
+/// Structure of a PCID. A PCID has to be <= 4096 for x86_64.
+pub const Pcid = packed struct {
+    value: u12,
+
+    /// Create a new PCID
+    pub fn init(pcid: u12) Pcid {
+        return .{
+            .value = pcid,
+        };
+    }
+};
+
+// ~~~ [ zig-bitjuggle/bitjuggle.zig ] ~~~
+
+/// Obtains the `number_of_bits` bits starting at `start_bit`
+/// Where `start_bit` is the lowest significant bit to fetch
+///
+/// ```zig
+/// const a: u8 = 0b01101100;
+/// const b = getBits(a, 2, 4);
+/// try testing.expectEqual(@as(u4,0b1011), b);
+/// ```
+pub fn getBits(target: anytype, comptime start_bit: comptime_int, comptime number_of_bits: comptime_int) std.meta.Int(.unsigned, number_of_bits) {
```

Comment:
What does `comptime` mean before a function argument? Also, what does `std.meta.Int(.unsigned, number_of_bits)` do?

Crazy that a return type can depend on the function argument (dependent types!).

What does `.unsigned` mean?

Reply:
> What does comptime mean before a function argument?

ref: https://ziglang.org/documentation/master/#comptime

>  A `comptime` parameter means that: 
> * At the callsite, the value must be known at compile-time, or it is a compile error.
> * In the function definition, the value is known at compile-time.

ref: https://ziglang.org/documentation/master/#Primitive-Types

`comptime_int`: Only allowed for [comptime](https://ziglang.org/documentation/master/#comptime)-known values. The type of integer literals.

> Also, what does std.meta.Int(.unsigned, number_of_bits) do?

`std.meta.Int(.unsigned, number_of_bits)` is a haxxy way to specify the type `uXXX` where XXX is defined by the value of the compile-time known integer literal passed as `number_of_bits` argument to `getBits`.